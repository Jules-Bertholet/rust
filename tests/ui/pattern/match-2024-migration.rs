//@ run-pass
//@ run-rustfix
//@ rustfix-only-machine-applicable
#![allow(unused_variables, unused_mut)]
#![warn(dereferencing_mut_binding)]
fn main() {
    // A tuple is a "non-reference pattern".
    // A `mut` binding pattern resets the binding mode to by-value
    // in edition <= 2021.

    let mut p = (0u8, 0u8);
    let (a, mut b) = &mut p;
    //~^ WARN dereferencing `mut`
    //~| WARN this changes meaning in Rust 2024
    let _: u8 = b;

    let (a, mut b) = &p;
    //~^ WARN dereferencing `mut`
    //~| WARN this changes meaning in Rust 2024
    let _: u8 = b;

    let (a, mut b @ _) = &mut p;
    //~^ WARN dereferencing `mut`
    //~| WARN this changes meaning in Rust 2024
    let _: u8 = b;

    let (a, mut b @ _) = &p;
    //~^ WARN dereferencing `mut`
    //~| WARN this changes meaning in Rust 2024
    let _: u8 = b;

    let mut p = (&0u8, &0u8);
    let (a, mut b) = &mut p;
    //~^ WARN dereferencing `mut`
    //~| WARN this changes meaning in Rust 2024
    let _: &u8 = b;

    let (a, mut b) = &p;
    //~^ WARN dereferencing `mut`
    //~| WARN this changes meaning in Rust 2024
    let _: &u8 = b;

    let (a, mut b @ _) = &mut p;
    //~^ WARN dereferencing `mut`
    //~| WARN this changes meaning in Rust 2024
    let _: &u8 = b;

    let (a, mut b @ _) = &p;
    //~^ WARN dereferencing `mut`
    //~| WARN this changes meaning in Rust 2024
    let _: &u8 = b;
}
