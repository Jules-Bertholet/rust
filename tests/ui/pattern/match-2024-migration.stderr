warning: dereferencing `mut` binding
  --> $DIR/match-2024-migration.rs:12:13
   |
LL |     let (a, mut b) = &mut p;
   |             ^^^^^
   |             |
   |             `mut` dereferences the type of this binding
   |             help: desugar the match ergonomics: `&mut mut b`
   |
   = warning: this changes meaning in Rust 2024
   = note: for more information, see FIXME none yet
help: this will change in edition 2024
  --> $DIR/match-2024-migration.rs:12:13
   |
LL |     let (a, mut b) = &mut p;
   |             ^^^^^
note: the lint level is defined here
  --> $DIR/match-2024-migration.rs:5:9
   |
LL | #![warn(dereferencing_mut_binding)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^

warning: dereferencing `mut` binding
  --> $DIR/match-2024-migration.rs:17:13
   |
LL |     let (a, mut b) = &p;
   |             ^^^^^
   |             |
   |             `mut` dereferences the type of this binding
   |             help: desugar the match ergonomics: `&(mut b)`
   |
   = warning: this changes meaning in Rust 2024
   = note: for more information, see FIXME none yet
help: this will change in edition 2024
  --> $DIR/match-2024-migration.rs:17:13
   |
LL |     let (a, mut b) = &p;
   |             ^^^^^

warning: dereferencing `mut` binding
  --> $DIR/match-2024-migration.rs:22:13
   |
LL |     let (a, mut b @ _) = &mut p;
   |             --------^
   |             |
   |             `mut` dereferences the type of this binding
   |             help: desugar the match ergonomics: `&mut mut b @`
   |
   = warning: this changes meaning in Rust 2024
   = note: for more information, see FIXME none yet
help: this will change in edition 2024
  --> $DIR/match-2024-migration.rs:22:13
   |
LL |     let (a, mut b @ _) = &mut p;
   |             ^^^^^^^^^

warning: dereferencing `mut` binding
  --> $DIR/match-2024-migration.rs:27:13
   |
LL |     let (a, mut b @ _) = &p;
   |             ^^^^^^^^^ `mut` dereferences the type of this binding
   |
   = warning: this changes meaning in Rust 2024
   = note: for more information, see FIXME none yet
help: this will change in edition 2024
  --> $DIR/match-2024-migration.rs:27:13
   |
LL |     let (a, mut b @ _) = &p;
   |             ^^^^^^^^^
help: desugar the match ergonomics
   |
LL |     let (a, &(mut b @ _)) = &p;
   |             ~~~~~~~~~  +

warning: dereferencing `mut` binding
  --> $DIR/match-2024-migration.rs:33:13
   |
LL |     let (a, mut b) = &mut p;
   |             ^^^^^
   |             |
   |             `mut` dereferences the type of this binding
   |             help: desugar the match ergonomics: `&mut mut b`
   |
   = warning: this changes meaning in Rust 2024
   = note: for more information, see FIXME none yet
help: this will change in edition 2024
  --> $DIR/match-2024-migration.rs:33:13
   |
LL |     let (a, mut b) = &mut p;
   |             ^^^^^

warning: dereferencing `mut` binding
  --> $DIR/match-2024-migration.rs:38:13
   |
LL |     let (a, mut b) = &p;
   |             ^^^^^
   |             |
   |             `mut` dereferences the type of this binding
   |             help: desugar the match ergonomics: `&(mut ref b)`
   |
   = warning: this changes meaning in Rust 2024
   = note: for more information, see FIXME none yet
help: this will change in edition 2024
  --> $DIR/match-2024-migration.rs:38:13
   |
LL |     let (a, mut b) = &p;
   |             ^^^^^

warning: dereferencing `mut` binding
  --> $DIR/match-2024-migration.rs:43:13
   |
LL |     let (a, mut b @ _) = &mut p;
   |             --------^
   |             |
   |             `mut` dereferences the type of this binding
   |             help: desugar the match ergonomics: `&mut mut b @`
   |
   = warning: this changes meaning in Rust 2024
   = note: for more information, see FIXME none yet
help: this will change in edition 2024
  --> $DIR/match-2024-migration.rs:43:13
   |
LL |     let (a, mut b @ _) = &mut p;
   |             ^^^^^^^^^

warning: dereferencing `mut` binding
  --> $DIR/match-2024-migration.rs:48:13
   |
LL |     let (a, mut b @ _) = &p;
   |             ^^^^^^^^^ `mut` dereferences the type of this binding
   |
   = warning: this changes meaning in Rust 2024
   = note: for more information, see FIXME none yet
help: this will change in edition 2024
  --> $DIR/match-2024-migration.rs:48:13
   |
LL |     let (a, mut b @ _) = &p;
   |             ^^^^^^^^^
help: desugar the match ergonomics
   |
LL |     let (a, &(mut ref b @ _)) = &p;
   |             ~~~~~~~~~~~~~  +

warning: 8 warnings emitted

